File: tools/simulate.ts
import { createHash } from "node:crypto";
import path from "node:path";

import { loadContent } from "../core/src/content/loader";
import { AutopickPolicy, step } from "../core/src/sim/engine";
import { formatLogEntry } from "../core/src/sim/log";
import { createInitialRunnerState } from "../core/src/sim/state";

interface CliArgs {
  seed: number | string;
  steps: number;
  biome?: string;
  autopick: AutopickPolicy;
}

function parseArgs(argv: string[]): CliArgs {
  const parsed: CliArgs = {
    seed: 1337,
    steps: 30,
    autopick: "safe"
  };

  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (token === "--seed") {
      const next = argv[index + 1];
      if (!next) {
        throw new Error("--seed requires a value.");
      }
      const numeric = Number(next);
      parsed.seed = Number.isFinite(numeric) && !Number.isNaN(numeric) ? numeric : next;
      index += 1;
      continue;
    }
    if (token === "--steps") {
      const next = argv[index + 1];
      if (!next) {
        throw new Error("--steps requires a value.");
      }
      const numeric = Number(next);
      if (!Number.isInteger(numeric) || numeric <= 0) {
        throw new Error(`--steps must be a positive integer. Received: ${next}`);
      }
      parsed.steps = numeric;
      index += 1;
      continue;
    }
    if (token === "--biome") {
      const next = argv[index + 1];
      if (!next) {
        throw new Error("--biome requires a biome id.");
      }
      parsed.biome = next;
      index += 1;
      continue;
    }
    if (token === "--autopick") {
      const next = argv[index + 1];
      if (!next || !["safe", "random", "greedy"].includes(next)) {
        throw new Error("--autopick must be one of: safe, random, greedy.");
      }
      parsed.autopick = next as AutopickPolicy;
      index += 1;
      continue;
    }
    throw new Error(`Unknown argument: ${token}`);
  }

  return parsed;
}

function deterministicSignature(
  payload: Record<string, unknown>,
  logLines: string[]
): string {
  return createHash("sha256")
    .update(JSON.stringify({ payload, logLines }))
    .digest("hex")
    .slice(0, 16);
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));
  const contentDir = path.resolve(process.cwd(), "content");
  const content = await loadContent(contentDir);

  const biomeId = args.biome ?? content.biomes[0]?.id;
  if (!biomeId) {
    throw new Error("No biome available in content.");
  }
  if (!content.biomeById[biomeId]) {
    throw new Error(`Unknown biome '${biomeId}'.`);
  }

  let state = createInitialRunnerState(args.seed, biomeId);
  const allLogs: string[] = [];

  for (let index = 0; index < args.steps; index += 1) {
    if (state.dead) {
      break;
    }
    const result = step(state, content, args.autopick);
    state = result.state;
    result.logEntries.forEach((entry) => {
      const line = formatLogEntry(entry);
      allLogs.push(line);
      // eslint-disable-next-line no-console
      console.log(line);
    });
  }

  const inventorySummary = Object.entries(state.inventory)
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([itemId, qty]) => `${itemId}x${qty}`)
    .join(", ");
  const flagsSummary = [...state.flags].sort().join(", ");

  // eslint-disable-next-line no-console
  console.log("\n--- Final Summary ---");
  // eslint-disable-next-line no-console
  console.log(`seed=${state.seed} autopick=${args.autopick} steps=${state.step}/${args.steps}`);
  // eslint-disable-next-line no-console
  console.log(`distance=${state.distance.toFixed(2)} time=${state.time} biome=${state.biomeId}`);
  // eslint-disable-next-line no-console
  console.log(
    `meters={stamina:${state.meters.stamina.toFixed(2)}, hydration:${state.meters.hydration.toFixed(
      2
    )}, morale:${state.meters.morale.toFixed(2)}} injury=${state.injury.toFixed(2)}`
  );
  // eslint-disable-next-line no-console
  console.log(`dead=${state.dead}${state.deathReason ? ` reason=${state.deathReason}` : ""}`);
  // eslint-disable-next-line no-console
  console.log(`flags=[${flagsSummary}]`);
  // eslint-disable-next-line no-console
  console.log(`inventory=[${inventorySummary}]`);

  const signature = deterministicSignature(
    {
      seed: state.seed,
      autopick: args.autopick,
      distance: state.distance,
      step: state.step,
      dead: state.dead,
      deathReason: state.deathReason,
      meters: state.meters,
      injury: state.injury,
      biome: state.biomeId,
      flags: [...state.flags].sort(),
      inventory: Object.entries(state.inventory).sort((a, b) => a[0].localeCompare(b[0]))
    },
    allLogs
  );

  // eslint-disable-next-line no-console
  console.log(`signature=${signature}`);
}

void main().catch((error) => {
  // eslint-disable-next-line no-console
  console.error(error instanceof Error ? error.message : String(error));
  process.exitCode = 1;
});

File: bit_life_survival/tools/simulate.py
from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Literal

import typer
from rich.console import Console
from rich.table import Table

from bit_life_survival.core.engine import create_initial_state, run_simulation
from bit_life_survival.core.loader import ContentValidationError, load_content

app = typer.Typer(add_completion=False, help="Run deterministic headless simulation for balancing and testing.")
console = Console()

AutopickPolicy = Literal["safe", "random", "greedy"]


def _normalize_seed(raw_seed: str) -> int | str:
    try:
        return int(raw_seed)
    except ValueError:
        return raw_seed


def _state_signature_payload(state, logs) -> dict:
    return {
        "seed": state.seed,
        "step": state.step,
        "distance": round(state.distance, 6),
        "time": state.time,
        "biome_id": state.biome_id,
        "meters": {
            "stamina": round(state.meters.stamina, 6),
            "hydration": round(state.meters.hydration, 6),
            "morale": round(state.meters.morale, 6),
        },
        "injury": round(state.injury, 6),
        "flags": sorted(state.flags),
        "inventory": dict(sorted(state.inventory.items())),
        "equipped": state.equipped.model_dump(mode="python"),
        "dead": state.dead,
        "death_reason": state.death_reason,
        "event_cooldowns": dict(sorted(state.event_cooldowns.items())),
        "rng_state": state.rng_state,
        "rng_calls": state.rng_calls,
        "timeline": [entry.to_dict() for entry in logs],
    }


@app.command()
def main(
    seed: str = typer.Option("123", "--seed", help="Seed value (int or string)."),
    steps: int = typer.Option(50, "--steps", min=1, help="Max number of simulation steps."),
    biome: str = typer.Option("suburbs", "--biome", help="Starting biome id."),
    autopick: AutopickPolicy = typer.Option("safe", "--autopick", help="Choice policy: safe|random|greedy."),
) -> None:
    content_dir = Path(__file__).resolve().parents[1] / "content"
    try:
        content = load_content(content_dir)
    except ContentValidationError as exc:
        console.print(f"[bold red]Content load failed:[/bold red] {exc}")
        raise typer.Exit(1) from exc

    if biome not in content.biome_by_id:
        console.print(f"[bold red]Unknown biome '{biome}'.[/bold red]")
        raise typer.Exit(1)

    state = create_initial_state(_normalize_seed(seed), biome)
    final_state, logs = run_simulation(state, content, steps=steps, policy=autopick)

    for entry in logs:
        console.print(entry.format(), markup=False)

    summary = Table(title="Simulation Summary")
    summary.add_column("Field", style="cyan", no_wrap=True)
    summary.add_column("Value", style="white")
    summary.add_row("Seed", str(final_state.seed))
    summary.add_row("Policy", autopick)
    summary.add_row("Steps", f"{final_state.step}/{steps}")
    summary.add_row("Distance", f"{final_state.distance:.2f}")
    summary.add_row("Time", str(final_state.time))
    summary.add_row(
        "Meters",
        (
            f"stamina={final_state.meters.stamina:.2f}, "
            f"hydration={final_state.meters.hydration:.2f}, "
            f"morale={final_state.meters.morale:.2f}"
        ),
    )
    summary.add_row("Injury", f"{final_state.injury:.2f}")
    summary.add_row("Dead", str(final_state.dead))
    summary.add_row("Death Reason", final_state.death_reason or "-")
    summary.add_row("Flags", ", ".join(sorted(final_state.flags)) or "-")
    summary.add_row(
        "Inventory",
        ", ".join(f"{item_id}x{qty}" for item_id, qty in sorted(final_state.inventory.items())) or "-",
    )
    console.print()
    console.print(summary)

    payload = _state_signature_payload(final_state, logs)
    signature = hashlib.sha256(json.dumps(payload, sort_keys=True).encode("utf-8")).hexdigest()[:16]
    console.print(f"\n[bold green]Deterministic signature:[/bold green] {signature}")


if __name__ == "__main__":
    app()
